### DISCUSSION TITLE
Is there a better way to store/pass a reference to "self" within a service?

### DISCUSSION THREAD

**Discussion by Asderex** [statelyai/xstate#1573] [Q&A] [answered]
Hi, Is there any way a service can obtain a reference to itself from inside the service? I was going to pass it in as an event and store it on the context after the service was started but was wondering if there was a more elegant solution. I'm sure there are anti-pattern alarms ringing **loudly** but this solution would be ideal for my particular problem and passing the reference in as an event after the service has started feels unnecesarily janky... ideally i would like to be able to do something like....

`actions: (context, event) => event.someOtherMachineReference.send(
    {
   type: 'REGISTER_FOR_UPDATES', 
returnAddress: context.self
}
)
`

where context.self is a reference to the service running the action and someOtherMachineReference will use the passed in reference (self.send) to send future updates.  (note: for the avoidance of doubt I'm aware of onTransition)

"No" is both a complete sentence and welcome reponse if correct.

**Comment by Andarist** [comment, marked as answer]
> "No" is both a complete sentence and welcome reponse if correct.

😅 

No.

**Reply by Andarist** [reply]
But to be more serious - there probably should be a better way to do this. Could u file an issue asking about this so we can discuss more details there? This probably would not be available as part of the `context` as that is fully under the control of the user and a `self`/`sessionId` is a system-controlled "variable" and as such we probably should provide it as part of the 3rd argument (so-called `meta`). This ain't super trivial to add right now because of the Machine/interpreter APIs split and how they interact with each other - but we should brainstorm how to make this happen in the least disruptive manner.

**Reply by Asderex** [reply]
Thanks for the confirmation @Andarist - looks like this is something David has been thinking about too so will file that issue if he wants to track it expliciltly.

**Comment by davidkpiano** [comment]
In version 5, there will be. Being able to reference `self` is important in an actor-model framework (Akka provides the "self" value in the "actor context") and there's even a way to do this for `assign` calls in the v5 branch:

```js
actions: assign((ctx, e, { spawn, self }) => {
  // `self` refers to the ActorRef instance of this machine
})
```

We just need to make that `self` reference generally available to all "meta data".

**Reply by Asderex** [reply]
Thanks David - It sounds like this is well in hand but did you want me to file an issue as suggested by @Andarist for tracking purposes or is this not required?

**Reply by davidkpiano** [reply]
It's not required.
